### DISCUSSION TITLE
How to run a shell function with execute or become

### DISCUSSION THREAD

**Discussion by rbhanot4739** [junegunn/fzf#4016] [Q&A]
Lets say we have a shell function defined which can be do different things depending on the arguments passed to it. How can run this function by binding a key with `become` or `execute` action ? Here is the example code

```sh
greet(){
echo "Hello $@"
}

ls | fzf --bind '<some-key>:become(greet {})
```
Unfortunately the above does not seem to work key press fails with `command greet not found`. I have also tried exporting the function but with no success. 
One workaround is to simply run the command directly inside the `become` like `become(echo "Hello " {}` which works but then if I have logic that I want to reuse across multiple bindings, I would have to repeat the same code. 

Another way is to put this function into its own script and then invoke that script, which works, but moving these functions into their own scripts can be painful and overkill if I am working on something more involved .

Am I missing something obvious here, I have searched through docs, example sections, advanced section but could not find anything.

**Comment by LangLangBart** [comment]
This will depend on your shell. If you are using `bash`, try it with `export -f`. As a good measure, assign the `SHELL` variable to your `bash` executable, Alternatively, there is also the `--with-shell` flag.


```sh
greet(){
echo "Hello $@"
}
export -f greet
ls | SHELL=$(which bash) fzf --bind '<some-key>:become(greet {})
```

---

> I have also tried exporting the function but with no success.

If you are using `zsh`, there is no `export -f`. See this discussion for workarounds: #3993



---

The key lesson to take away is that `fzf` executes these functions in a child process like this:

> [!NOTE]
> _fzf runs the command with `$SHELL -c` if `SHELL` is set, otherwise with `sh -c`_

**Comment by rbhanot4739** [comment]
@LangLangBart Yes, I am using `zsh`. I had considered squishing the function into a single command string but that can become really hard to read and debug.  For shorter and one liner functions this approach would work just fine however for longer and complicated functions its non-trivial and I think maybe `.zshenv` is the way but as you mentioned that has another side effect of very loose code locality !

Is it possible to run the zsh function inside a bash subshell when fzf executes it ? I tried running the following from zshell but it din't work
```
greet(){
echo "Hello $@"
}
export -f greet
ls | SHELL=$(which bash) fzf --bind '<some-key>:become(greet {})
```

**Reply by LangLangBart** [reply]
Within a `zsh` environment, you can't run `export -f greet`; it will throw an error.

In a `bash` environment, like in a script, it will work.
```sh
#!/usr/bin/env bash
greet(){
echo "Hello $@"
}
export -f greet
ls | SHELL=$(which bash) fzf --bind '<some-key>:become(greet {})
```

---

Alternatively, source the function from a specific file and then run it.


```sh
# file: ~/greet.zsh
greet(){
echo "Hello $@"
}

# command line
ls | fzf --bind '<some-key>:become(source ~/greet.zsh; greet {})
```

---

There is also `autoload`, but this would also require you to write the function in its own file 
```sh
man zshmisc | less --pattern "AUTOLOADING FUNCTIONS"
```

This is just an example of using the `is-at-least`[^1] function that comes with every `zsh` version. Use `autoload` to call it and test if your `fzf` version is at least `0.40.1`. This is just a demo to illustrate the point.

- `is-at-least` check if a given command version meets the requirements
```sh
autoload -Uz is-at-least
is-at-least 0.40.1 $(fzf --version) && echo YES || echo NO 
```

There are several useful functions available with every `zsh` version. You can find them in the `share/zsh/functions` directory of your zsh distribution, along with many other completion functions. `zsh` even includes a `tetris` game, which might not be very helpful, but is amusing. 

```sh
man zshcontrib | less --pattern tetris

echo "Play Tetris" | SHELL="$(which zsh)" fzf --bind 'enter:become:autoload -Uz tetriscurses; tetriscurses'
```

[^1]: [zsh/code Functions/Misc/is-at-least](https://sourceforge.net/p/zsh/code/ci/master/tree/Functions/Misc/is-at-least)

**Comment by rbhanot4739** [comment]
This seems to be the most convenient way if I put this inside a script like `test.sh`.

```
#!/usr/bin/env bash
greet(){
echo "Hello $@"
}
export -f greet
ls | SHELL=$(which bash) fzf --bind '<some-key>:become(greet {})
```
 But it still does not work if I run this script like `bash test.sh`.

Even with autoloading the function fzf somehow fails to correctly load the function, here is the snippet from `.zshrc`
```
fpath=(~/zsh_functions $fpath);
autoload -U $fpath[1]/*(.:t)
```
Now if I open a new shell and run `greet Foo` I get `Hello Foo`, but if I do
```
❯ ls |SHELL="$(which zsh)" fzf --bind 'enter:become(greet)'
zsh:1: command not found: greet
```
Even with explicit autoloading it isn't working.
```
❯ ls |SHELL="$(which zsh)" fzf --bind 'enter:become:autoload -Uz greet; greet'
zsh:1: greet: function definition file not found
```
Here is the functions directory and file named `greet` inside it.
```
zsh_functions
└── greet
```

**Reply by LangLangBart** [reply]
> But it still does not work if I run this script like `bash test.sh`.

Unable to reproduce the issue. Tried it with the following.

```sh
cat <<'EOF' >/tmp/test.sh
#!/usr/bin/env bash
greet(){
echo "Hello $@"
}
export -f greet
ls | SHELL=$(which bash) fzf --bind 'enter:become(greet {})'
EOF

bash /tmp/test.sh
```

Could you enable `xtrace` ?
```sh
bash -x /tmp/test.sh
```

---

> Even with explicit autoloading it isn't working.

Did you restart your shell before running the command to ensure that the `FPATH` variable was properly exported? Check the output when printing the listed paths in `fpath`.

```sh
: | SHELL="$(which zsh)" fzf --bind 'start:become: echo "$FPATH"'
```

or try 
```sh
fpath=(~/zsh_functions $fpath)
export FPATH=$FPATH
```

Testing it in a clean environment:
```sh
command env -i HOME=$HOME TERM=$TERM USER=$USER PATH=$PATH zsh -f
cd ~
mkdir zsh_functions

cat <<'EOF' > zsh_functions/greet
echo "Hello $@"
EOF

fpath=(~/zsh_functions $fpath)
export FPATH=$FPATH

ls | SHELL="$(which zsh)" fzf --bind 'enter:become:autoload -Uz greet; greet {}'
Hello ...
```

**Comment by rbhanot4739** [comment]
@LangLangBart  I did some more testing with bash script method and noticed something really strange, the script works outside of tmux but not within tmux. Here is the example
```sh
❯ cat  /tmp/test.sh
#!/usr/bin/env bash
greet() {
  echo "Hello $@"
}
export -f greet
ls | SHELL=$(which bash) fzf --bind 'enter:become(greet {})'

❯ FZF_DEFAULT_OPTS=""
❯ bash /tmp/test.sh
Hello catpuccin.txt
```

Now if I add the `--tmux` option to default opts, it breaks
```sh
❯ cat /tmp/test.sh
#!/usr/bin/env bash
greet() {
  echo "Hello $@"
}
export -f greet
ls | SHELL=$(which bash) fzf --bind 'enter:become(greet {})'

❯ FZF_DEFAULT_OPTS="--tmux"
❯ bash /tmp/test.sh
/opt/homebrew/bin/bash: line 1: greet: command not found
```
Not sure what's going on here.

**Reply by LangLangBart** [reply]
> the script works outside of tmux but not within tmux

See https://github.com/junegunn/fzf/issues/4001#issuecomment-2367823867

**Comment by rbhanot4739** [comment]
@LangLangBart Sorry for stretching this, but I noticed another issue.

```sh
tmux ls -F '#{session_name}' | fzf --bind 'enter:become(tmux attach -t {})'
open terminal failed: can't use /dev/tty
```
It only happens when the command is run as fzf `action`, if I do `tmux attach -t $selection` directly after `selection=$(tmux ls -F '#{session_name}' | fzf)`  the session is attached because I guess the command runs in an interactive shell where there is a tty available. I tried passing explicitly as well
```sh
tmux ls -F '#{session_name}' | SHELL=`which bash` fzf --bind 'enter:become(tmux attach -t {})'
open terminal failed: can't use /dev/tty
```

I assume this to be a known behavior but could not find any discussion/issues related to this.

Also did you try running below from within a tmux session ? For me, it works outside of tmux but not from within tmux ? Not sure if I am missing something here but running above from within tmux throws the same `command greet not found`.

> Unable to reproduce the issue. Tried it with the following:.
> 
> ```
> cat <<'EOF' >/tmp/test.sh
> #!/usr/bin/env bash
> greet(){
> echo "Hello $@"
> }
> export -f greet
> ls | SHELL=$(which bash) fzf --bind 'enter:become(greet {})'
> EOF
> 
> bash /tmp/test.sh
> ```
> Could you enable xtrace ?
> 
> `bash -x /tmp/test.sh`

**Reply by LangLangBart** [reply]
> ```shell
> tmux ls -F '#{session_name}' | fzf --bind 'enter:become:tmux attach -t {}'
> open terminal failed: can't use /dev/tty
> ```


Have you tried the suggestions in the wiki[^1] instead ?



---

> For me, it works outside of tmux but not from within tmux ?


Running it with `--no-tmux` should work, both inside and outside of `tmux`.
```sh
ls | SHELL=$(which bash) fzf --no-tmux --bind 'enter:become(greet {})'
```

If you want to use the `--tmux` flag, you will need to use the current development version. See commit 4161403.

See `BUILD.md`[^2] for instructions if you would like to test it, and thank the maintainer for fixing it.

[^1]: [Examples tmux · junegunn/fzf Wiki · GitHub](https://github.com/junegunn/fzf/wiki/Examples#tmux)
[^2]: [fzf/BUILD.md at master · junegunn/fzf · GitHub](https://github.com/junegunn/fzf/blob/master/BUILD.md)

**Comment by rbhanot4739** [comment]
> Have you tried the suggestions in the wiki[1](https://github.com/junegunn/fzf/discussions/4016#user-content-fn-1-e6e46cbe27eda5ff2273cce6f501e339) instead ?

This regular `attach-sssion` works just fine, as I mentioned that if I get the selection from fzf like below it works fine.
```
selection=$(tmux ls -F '#{session_name}' | fzf)
tmux attach-session -t $selection
```

But when I do `tmux ls -F '#{session_name}' | fzf --bind 'enter:become(tmux attach -t {})'` it doesn't. Infact if I bound a function to key with `--bind` which internally calls `tmux attach-session` it still fails with the same error `open terminal failed: can't use /dev/tty`.

**Reply by LangLangBart** [reply]
> open terminal failed: can't use /dev/tty

The error message sits in `tmux`[^1][^2] and is triggered by this line:

```c
if (strcmp(c->ttyname, ttynam) == 0)
```

`ttynam` derives its value from `_PATH_TTY`[^3] (if not otherwise defined it is `/dev/tty`)
`c->ttyname` gets its value from `ttyname(STDIN_FILENO)`[^4]

Checking the names of the file descriptors (stdin, stdout, stderr) through `fzf's` `become` action.
```sh
fzf --bind 'start:become:lsof -p $$ -a -d 0-2'
# COMMAND  PID  USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
# lsof    1584 paria    0r   CHR    2,0      0t0  315 /dev/tty
# lsof    1584 paria    1u   CHR  16,44  0t13551 1297 /dev/ttys044
# lsof    1584 paria    2u   CHR  16,44  0t13551 1297 /dev/ttys044

# or using the `JXA-ObjC Bridge` (macOS)
fzf --with-shell 'osascript -l JavaScript -e' \
  --bind 'start:become:ObjC.import("unistd"); [$.ttyname($.STDIN_FILENO), $.ttyname($.STDOUT_FILENO), $.ttyname($.STDERR_FILENO)].join("\n")'
# /dev/tty
# /dev/ttys044
# /dev/ttys044
```

If we now require `stdin` to point to the same location as `stderr` or `stdout`, it might just work[^5].
```sh
fzf --bind 'start:become:lsof -p $$ -a -d 0-2 </dev/stderr'
# COMMAND  PID  USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
# lsof    1666 paria    0u   CHR  16,44  0t14946 1297 /dev/ttys044
# lsof    1666 paria    1u   CHR  16,44  0t14946 1297 /dev/ttys044
# lsof    1666 paria    2u   CHR  16,44  0t14946 1297 /dev/ttys044
```

@junegunn would you approve of that workaround?
```sh
tmux ls -F '#{session_name}' | fzf --bind 'enter:become:tmux attach -t {} </dev/stderr'
```



[^1]: [tmux/server-client.c at 826ba515beef76dff7d8865599a19e06ae1c4f6a · tmux/tmux · GitHub](https://github.com/tmux/tmux/blob/826ba515beef76dff7d8865599a19e06ae1c4f6a/server-client.c#L320)
[^2]: [tmux/cmd-attach-session.c at 826ba515beef76dff7d8865599a19e06ae1c4f6a · tmux/tmux · GitHub](https://github.com/tmux/tmux/blob/826ba515beef76dff7d8865599a19e06ae1c4f6a/cmd-attach-session.c#L133)
[^3]: [tmux/compat.h at 826ba515beef76dff7d8865599a19e06ae1c4f6a · tmux/tmux · GitHub](https://github.com/tmux/tmux/blob/826ba515beef76dff7d8865599a19e06ae1c4f6a/compat.h#L110)
[^4]: [tmux/client.c at 826ba515beef76dff7d8865599a19e06ae1c4f6a · tmux/tmux · GitHub](https://github.com/tmux/tmux/blob/826ba515beef76dff7d8865599a19e06ae1c4f6a/client.c#L306)
[^5]: [bash - shell command `tmux` throws `can't use /dev/tty` error - Unix &amp; Linux Stack Exchange](https://unix.stackexchange.com/questions/404173)

**Comment by bdjnk** [comment]
One straightforward way to retain locality of behavior relies on a conditional switch within one file, either to search or to run the bound function, as demonstrated in this `test.sh` file.

```
greet() {
	echo "Hello $@"
}

find() {
	ls | fzf --bind 'h:become(./test.sh greet {})'
}

case $1 in
	greet ) greet $2 ;;
	*     ) find     ;;
esac
```

The result will be simple to understand and flexible enough to handle complex scenarios as needed.
